docs/lang/articles/basic/external.md

```diff
@@ -4,197 +4,195 @@ sidebar_position: 5
+We use NumPy arrays as an example to illustrate the data transfer process because NumPy arrays are the most commonly used external arrays in Taichi. The same steps apply to PyTorch tensors and Paddle tensors.
 
-```python {8}
-@ti.kernel
-def my_kernel():
-    for i in x:
-        x[i] = i * 2
-
-x = ti.field(ti.f32, 4)
-my_kernel()
-x_np = x.to_numpy()
-print(x_np)  # np.array([0, 2, 4, 6])
-```
+There are two ways to import a NumPy array `arr` to the Taichi scope:
 
-**Import data from NumPy arrays to Taichi fields** via `from_numpy()`. This allows us to initialize Taichi fields via NumPy arrays:
+1. Create a Taichi field `f` whose shape and dtype match the shape and dtype of `arr`, and call the method `f.from_numpy(arr)` to copy the data in `arr` into `f`. This approach is preferred when the original array is visited frequently from elsewhere in the Taichi scope (for example, in the case of texture sampling).
 
-```python {3}
-x = ti.field(ti.f32, 4)
-x_np = np.array([1, 7, 3, 5])
-x.from_numpy(x_np)
-print(x[0])  # 1
-print(x[1])  # 7
-print(x[2])  # 3
-print(x[3])  # 5
-```
+2. Pass `arr` as an argument to a kernel or a Taichi function using `ti.types.ndarray()` as the type hint. The argument is passed by reference without creating a copy of `arr`, and thus any modifications of this argument inside the kernel or Taichi function will also change the original array `arr`. This approach is preferred when the kernel or Taichi function that takes in the argument needs to manipulate the original array (for storage or filtering, for example).
+
+We now explain them in more details.
+
+## Data transfer between NumPy arrays and Taichi fields
+
+To import data from a NumPy array to a Taichi field, first make sure that the field and the array have the same shape:
 
-Likewise, Taichi fields can be **imported from and exported to PyTorch tensors**:
 ```python
-@ti.kernel
-def my_kernel():
-    for i in x:
-        x[i] = i * 2
-
-x = ti.field(ti.f32, 4)
-my_kernel()
-x_torch = x.to_torch()
-print(x_torch)  # torch.tensor([0, 2, 4, 6])
-
-x.from_torch(torch.tensor([1, 7, 3, 5]))
-print(x[0])  # 1
-print(x[1])  # 7
-print(x[2])  # 3
-print(x[3])  # 5
+x = ti.field(float, shape=(3, 3))
+a = np.arange(9).reshape(3, 3).astype(np.int32)
+x.from_numpy(a)
+print(x)
+#[[0 1 2]
+# [3 4 5]
+# [6 7 8]]
 ```
-And Taichi fields also can be **imported from and exported to Paddle tensors**:
+
+In the example above, the scalar field `x` and the array `a` have the same shape `(3, 3)`. This operation would fail if their shapes did not match. For vector and matrix fields, their shape matching with NumPy arrays is slightly different, which will be discussed in a later section.
+
+The field should also have the same dtype with the array; otherwise, an implicit type casting would occur - see [type system](../type_system/type.md).
```

Comment:
```suggestion
The field should also have the same dtype as the array; otherwise, an implicit type casting would occur - see [type system](../type_system/type.md).
```